### ISSUE TITLE
Inject custom taghelpers in Html.Raw()

### ISSUE THREAD

**Issue by pierre-weceipt** [aspnet/Mvc#4418]
Hi, I'm wondering if we can inject some new MVC and custom taghelpers in the razzor view with `@Html.raw()` and have them processed?

**Comment by Eilon** [comment]
@pierre-weceipt what exactly are you trying to do?

**Comment by pierre-weceipt** [comment]
I'm building a kind of "html Template editor". Templates that are then used to display content with data depending on the user (like graphs, etc.). 
I first thought of using some key-word in the Template and then switching it in my controller with html code containing the data but then thought of using taghelpers. As I'm using alloyeditor as a UI for the text parts of the template it is easy for me to add a button to insert some placeholders tag for the dynamic data (could be the taggelper). Then I'm passing the thing as a string to my view in a Html.Raw. It's working fine for regular html and I was wondering if it would work with taghelpers too (doubting it though).

TLDR : I want to inject custom taghelpers in my view as htmlstrings or strings passed to Html.Raw()

**Comment by Eilon** [comment]
@NTaylorMullen @dougbu any thoughts on this?

The thing with tag helpers is that they're intended to be *declarative*, and then ASP.NET/Razor calls them to do their work. The idea is that the developer writes some cool HTML-looking markup, gets awesome Intellisense and syntax highlighting, and then the tag helpers to go work and produce some output HTML.

It's certainly possible to "`new` up" a tag helper and get it to render some HTML, and then pass that to `@Html.Raw(...)`, but then a lot of the value of tag helpers is lost.

**Comment by dougbu** [comment]
- @pierre-weceipt do you want to use tag helper instances directly? Or is your idea to generate Razor code, run MVC's Razor view engine to generate C# and compile that, then execute the new class against a model? Or perhaps store generated HTML from tag helpers in a view model and use `@Html.Raw()` in static `.cshtml` files?
- An additional downside to passing anything through `@Html.Raw()` is performance. The `IHtmlContent` instances tag helpers manipulate avoid intermediate `string`s as much as possible for example.

**Comment by pierre-weceipt** [comment]
@Eilon : Taghelpers with highlighting and IntelliSense is really cool but just being able to convert a custom tag generated elsewhere in regular html is already a huge thing (at least for me).

@dougbu : I'm already storing HTML generated in my editor as a string and thought of including taghelpers to manage dynamic content. I wanted to then pass the string in a view model and use `@Html.Raw()` in static `.cshtml` files. This is why I was wondering if taghelpers would be processed by Razor if passed in `@Html.Raw()`. 

But I'm open to any suggestion

**Comment by Eilon** [comment]
@pierre-weceipt can you show what kind of syntax you're imagining? E.g. something like:

```
@Html.Raw(
    <some-tag-helper p1="v1" p2="v2">
        <a href="~/whatever">Link</a>
        <another-tag p3="v3" />
    </some-tag-helper>
)
```

?

**Comment by pierre-weceipt** [comment]
@Eilon Exactly, except that the following part would be passed through the model
```
<some-tag-helper p1="v1" p2="v2">
    <a href="~/whatever">Link</a>
    <another-tag p3="v3" />
</some-tag-helper>
```

**Comment by Eilon** [comment]
Sorry, what do you mean by "passed through the model"?

**Comment by pierre-weceipt** [comment]
I would have Something like this

A viewmodel :
```c#
public class MyViewModel {
   public string SomeHtmlContent { get; set; }
}
```
A controller :

```c#
public IActionResult Index()
        {
            var model= new MyViewModel ();
            model.SomeHtmlContent = '<some-tag-helper p1="v1" p2="v2"><a href="~/whatever">Link</a><another-tag p3="v3" /></some-tag-helper>'
            return View(model);
        }
```

And the view 

```c#
@model MyViewModel

@Html.Raw(MyViewModel.SomeHtmlContent);
```

**Comment by NTaylorMullen** [comment]
@pierre-weceipt we currently do not have any "nice" way to perform dynamic TagHelperification of output. `TagHelper`s are typically a compile time feature. If you'd like to build snippets and pass them to views I'd recommend throwing that HTML content into a partial, that way you'll be able to get all the `TagHelper` greatness.

If you absolutely must have a string that you then want to activate `TagHelper`s in you could run that string through the [MvcRazorHost as its own content stream](https://github.com/aspnet/Mvc/blob/dfe35bbef7211df567b26ce2a5b4a12e039226b9/src/Microsoft.AspNetCore.Mvc.Razor.Host/MvcRazorHost.cs#L260-L265) and then running the compiled content through the [DefaultRoslynCompilationService](https://github.com/aspnet/Mvc/blob/dfe35bbef7211df567b26ce2a5b4a12e039226b9/src/Microsoft.AspNetCore.Mvc.Razor/Internal/DefaultRoslynCompilationService.cs#L90). Both of which can be accessed from DI via [IMvcRazorHost](https://github.com/aspnet/Mvc/blob/dev/src/Microsoft.AspNetCore.Mvc.Razor.Host/IMvcRazorHost.cs) and [ICompilationService](https://github.com/aspnet/Mvc/blob/dev/src/Microsoft.AspNetCore.Mvc.Razor/Compilation/ICompilationService.cs). I do want to warn that this approach does not perform well.

**Comment by pierre-weceipt** [comment]
@NTaylorMullen the thing is that I'm receiving some content edited with a WYSIWIG editor with custom buttons to add some custom markers (html-like tags for ex) where I need to change things at display time. It can be different actions like :
- adding some specific html with fresh datas (charts for ex),
- display only a portion of the content (conditionnal processing);

Then I'm saving this html as a string and using it later to display content depending on the user. 
I thought of using the taghelpers to avoid the hasle of searching through the string for this markers in the controller and process their results. 

I think I will do it like this in a first time and then look into something more elegant if there's a performance issue although if you have a better idea I'll take it.

**Comment by Eilon** [comment]
BTW I would strongly caution you to consider security implications of allowing a client-side editor *within an app* to edit server code. Some tag helpers could have huge behavior implications, so allowing editing them while an app is running is potentially risky.

**Comment by pierre-weceipt** [comment]
@Eilon I take the point but I clean the HTML from any scripts and will just detect some specific words to inject data and will then pass everything to the view. The code is not really "running" on the server.

Thanks to you for the answers 😄
